lib/phoenix/endpoint.ex

```diff
@@ -213,6 +213,13 @@ defmodule Phoenix.Endpoint do
 
     * `:log_access_url` - log the access url once the server boots
 
+    * `:cswsh_check` - configure the CSWSH (Cross-Site WebSocket Hijacking) check
+    mechanism for the endpoint. When `:crsf_token` (the default), a CRSF token is
+    used. This mechanism uses the session and is stateful. When `:origin`, the
```

Comment:
I am not 100% sure I agree with this definition. We actually check both `:csrf_token` and `:origin`, it depends on which are provided. The thing is, because Phoenix does not allow you to read anything from headers, we actually don't need to check anything, unless you want to read the session. So maybe we call this `session_validator: :csrf_token | :origin`? Although it is weird to restrict it exclusively to session. 🤔  Thoughts?

Reply:
> We actually check both :csrf_token and :origin, it depends on which are provided.

We check the `origin` header unless there's no `origin` header in the request, that is if:
- the client is not a browser
- the browser is too old and doesn't support this header (which is rarely the case nowadays)
- for some reason the browser didn't send the header (very unlikely example: browser has a vulnerability which is used by a malicious actor)

CSRF token is always verified as soon as `:session` is set in `:connect_info`. Which makes this mechanism intrusive by setting the CSRF token in the page, turning Liveviews into non-cacheable pages.

> The thing is, because Phoenix does not allow you to read anything from headers, we actually don't need to check anything, unless you want to read the session.

Indeed, in this case only the `origin` header is checked, if it is configured to do so *and* is present in the request. I'm not sure it's even necessary for public Liveviews. Might be some other authentication mechanism (not Phoenix session) that benefit this though.

> So maybe we call this session_validator: :csrf_token | :origin?

I don't have a strong opinion on this. I feel like it's slightly different of session validation (which, we could argue, is about checking the cookie has not been tampered with or has not expired for example). Basically it's about checking the cookie comes from a domain we control. `Phoenix.Transports.LongPoll` uses the same mechanism, but I haven't double check yet that checking the origin is sufficient from a security perspective. But in this case indeed `cswsh` makes no sense.

 It makes me think a bit about [RFC8471 - The Token Binding Protocol Version 1.0](https://datatracker.ietf.org/doc/html/rfc8471). Maybe `:session_binding`?

> Although it is weird to restrict it exclusively to session.

Probably not, unless another custom session mechanism is in use.